### ISSUE TITLE
rc value in embedded default.lua code is nil when there is custom handler with custom exitcodes table defined, perhaps in other situations as well

### ISSUE THREAD

**Issue by etosan** [axkibe/lsyncd#372]
When one has config.exitcodes defined in custom script (like in lftp example), something, somewhere, get's confused and default.lua embedded in executable is getting (nil) instead of exitcodes table.

Offending lines I hacked-in nil check to get around this and make it to work are : 
https://github.com/axkibe/lsyncd/blob/master/default.lua#L133
https://github.com/axkibe/lsyncd/blob/master/default.lua#L165

When I return 'ok' on rc = (nil) script doesn't crash lsyncd on missing exit code table.
This efficiently disables all exitcodes checking in given situation, but it is sufficient for me to sidestep issue right now.

I was unable to find a place where exitcodes is lost, as code is quite advanced and hard to follow (at least for me).

This is certainly linked to #194 and maybe to others. 

It would be great if axkibe could look at it, with his insight into inner workings.

**Comment by axkibe** [comment]
Can you please provide, the exact Lsyncd version you're using, an example config script, and matching log files with -log all?

**Comment by etosan** [comment]
Very nice thanks for -log all suggestion, never would guessed it provides such detailed output.
Version used is latest clone from github.com, hash: 4740a2ca8fde64e9e05aa597681f12ced3b84312

Exact script dump - it's just very slightly tweaked original script:
```lua
--~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-- User configuration file for lsyncd.
--
--    Syncs with 'lftp'.
--
--~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

return {

	-----
	-- Spawns rsync for a list of events
	--
	action = function(inlet)

		-- gets all events ready for syncing
		local elist = inlet.getEvents(
			function(event)
				return event.etype ~= 'Init' and event.etype ~= 'Blanket'
			end
		)

		-----
		-- replaces filter rule by literals
		--
		local function sub(p)
			if not p then
				return
			end
			return p:gsub('%?', '\\?'):
			         gsub('%*', '\\*'):
			         gsub('%[', '\\['):
			         gsub('%]', '\\]')
		end

		local config = inlet.getConfig()

		local commands = elist.getPaths(

			function(etype, path1, path2)
				if etype == 'Delete' then
					if string.byte(path1, -1) == 47 then
                                                -- @eto: add exists check
						return 'rm -r '..
							config.targetdir..sub(path1)
					else
                                                -- @eto: add exists check
						return 'rm '..
							config.targetdir..sub(path1)

					end
				elseif
					etype == 'Create' or
					etype == 'Modify' or
		 			etype == 'Attrib'
		 		then
					if string.byte(path1, -1) == 47 then
						return 'mirror -R '..
							config.source..sub(path1)..' '..
							config.targetdir..sub(path1)
					else
						return 'put '..
							config.source..sub(path1)..
							' -o '..config.targetdir..sub(path1)
					end
				end
			end
		)

		if #commands == 0 then
			spawn(elist, '/bin/true')
			return
		end

		commands = table.concat(commands, ';\n')

		log('Normal', 'Calling lftp with commands\n', commands)

		spawn(elist, '/usr/bin/lftp',
			'<', commands,
			'-u', config.user..','..config.pass, config.host
		)
	end,

	-----
	-- Spawns the recursive startup sync
	--
	init = function(event)
		local config = event.config
		local inlet = event.inlet
		local excludes = inlet.getExcludes()
    local exclude_list = ''
    local includes = {}
    local include_list = ''
		local delete = nil

    if config.include then includes = config.include end

		if config.delete then delete = { '--delete', '--ignore-errors' } end

		if #excludes ~= 0 then
      exclude_list = ' --exclude-glob=\'' .. table.concat( excludes, '\' --exclude-glob=\'' ) .. '\''
		end

		if #includes~= 0 then
      include_list = ' --include-glob=\'' .. table.concat( includes, '\' --include-glob=\'' ) .. '\''
		end

		log('Normal', 'recursive startup lftp: ', config.source, ' to host: ', config.host)
    log('Normal', 'excluded list:' .. exclude_list)
    log('Normal', 'included list:' .. include_list)

		spawn(event, '/usr/bin/lftp',
			'-c',
			'open -u '..config.user..','..config.pass..' '..config.host..'; '..
			'mirror --reverse --parallel=6 --delete' .. exclude_list .. include_list .. ' ' ..config.source..' '..config.targetdir..';'
		)
	end,

	-----
	-- Checks the configuration.
	--
	prepare = function(config)

		if not config.host then
			error('lftps needs "host" configured', 4);
		end

		if not config.user then
			error('lftps needs "user" configured', 4);
		end

		if not config.pass then
			error('lftps needs "pass" configured', 4);
		end

		if not config.targetdir then
			error('lftp needs "targetdir" configured', 4)
		end

		if config.target then
			error('lftp needs NOT "target" configured', 4)
		end

--		if config.exclude then
--			error('lftp does not work with excludes', 4)
--		end

		if config.rsyncOpts then
			error('lftp needs NOT "rsyncOpts" configured', 4)
		end

		if string.sub(config.targetdir, -1) == '/' then
			error('please make targetdir not end with a /', 4)
		end

	end,

	-----
	-- Default delay
	--
	delay = 1,

  	-----
	-- Exit codes for rsync.
	--
	exitcodes = {
	  [0] = 'ok',
	  [1] = 'ok',
	},
}
```
Log dump contains too much information for my taste, so please give me more time, I will do some test setup and post dump later in a day or two at max (I hope).

**Comment by etosan** [comment]
Okay I created minimal setup with unaltered lsyncd version.

This is output capture:

```text
% ./lsyncd -log all test.lua
kernels clocks_per_sec=100
Call: configure( )
Exclude: toLuaPattern "ignore.file" = "/ignore%.file"
Inotify: inotify fd = 3
Call: initialize( )
10:32:14 Function: Inotify.addWatch( /tmp/lsyncd/data/ )
10:32:14 Inotify: addwatch( /tmp/lsyncd/data/ )-> 1 
10:32:14 Call: getAlarm( )
10:32:14 Alarm: runner.getAlarm returns: (true)
10:32:14 Masterloop: immediately handling delays.
10:32:14 Call: cycle( )
10:32:14 Function: invokeActions( "Sync1", (Timestamp: 4689867.9) )
10:32:14 Normal: recursive startup lftp: /tmp/lsyncd/data/ to host: sftp://fbsd-devbox
10:32:14 Normal: excluded list: --exclude-glob='ignore.file'
10:32:14 Normal: included list:
10:32:14 Exec: /usr/bin/lftp [-c] [open -u eto,ssh-agent sftp://fbsd-devbox; mirror --reverse --parallel=6 --delete --exclude-glob='ignore.file' /tmp/lsyncd/data/ /tmp/lsyncd/data;]
10:32:14 Function: write( (Timestamp: 4689867.9) )
10:32:14 Statusfile: writing now
10:32:14 Call: getAlarm( )
10:32:14 Alarm: runner.getAlarm returns: (false)
10:32:14 Masterloop: going into select ( no timeout )
10:32:14 Call: collectProcess( )          
10:32:14 Delay: collected an event
10:32:14 Error: Unknown exitcode "0" on startup of "/tmp/lsyncd/data/".
10:32:14 Error: Critical exitcode.
```
script lftp.lua

``` lua 
--~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-- User configuration file for lsyncd.
--
--    Syncs with 'lftp'.
--
--~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

return {

	-----
	-- Spawns rsync for a list of events
	--
	action = function(inlet)

		-- gets all events ready for syncing
		local elist = inlet.getEvents(
			function(event)
				return event.etype ~= 'Init' and event.etype ~= 'Blanket'
			end
		)

		-----
		-- replaces filter rule by literals
		--
		local function sub(p)
			if not p then
				return
			end
			return p:gsub('%?', '\\?'):
			         gsub('%*', '\\*'):
			         gsub('%[', '\\['):
			         gsub('%]', '\\]')
		end

		local config = inlet.getConfig()

		local commands = elist.getPaths(

			function(etype, path1, path2)
				if etype == 'Delete' then
					if string.byte(path1, -1) == 47 then
                                                -- @eto: add exists check
						return 'rm -r '..
							config.targetdir..sub(path1)
					else
                                                -- @eto: add exists check
						return 'rm '..
							config.targetdir..sub(path1)

					end
				elseif
					etype == 'Create' or
					etype == 'Modify' or
		 			etype == 'Attrib'
		 		then
					if string.byte(path1, -1) == 47 then
						return 'mirror -R '..
							config.source..sub(path1)..' '..
							config.targetdir..sub(path1)
					else
						return 'put '..
							config.source..sub(path1)..
							' -o '..config.targetdir..sub(path1)
					end
				end
			end
		)

		if #commands == 0 then
			spawn(elist, '/bin/true')
			return
		end

		commands = table.concat(commands, ';\n')

		log('Normal', 'Calling lftp with commands\n', commands)

		spawn(elist, '/usr/bin/lftp',
			'<', commands,
			'-u', config.user..','..config.pass, config.host
		)
	end,

	-----
	-- Spawns the recursive startup sync
	--
	init = function(event)
		local config = event.config
		local inlet = event.inlet
		local excludes = inlet.getExcludes()
    local exclude_list = ''
    local includes = {}
    local include_list = ''
		local delete = nil

    if config.include then includes = config.include end

		if config.delete then delete = { '--delete', '--ignore-errors' } end

		if #excludes ~= 0 then
      exclude_list = ' --exclude-glob=\'' .. table.concat( excludes, '\' --exclude-glob=\'' ) .. '\''
		end

		if #includes~= 0 then
      include_list = ' --include-glob=\'' .. table.concat( includes, '\' --include-glob=\'' ) .. '\''
		end

		log('Normal', 'recursive startup lftp: ', config.source, ' to host: ', config.host)
    log('Normal', 'excluded list:' .. exclude_list)
    log('Normal', 'included list:' .. include_list)

		spawn(event, '/usr/bin/lftp',
			'-c',
			'open -u '..config.user..','..config.pass..' '..config.host..'; '..
			'mirror --reverse --parallel=6 --delete' .. exclude_list .. include_list .. ' ' ..config.source..' '..config.targetdir..';'
		)
	end,

	-----
	-- Checks the configuration.
	--
	prepare = function(config)

		if not config.host then
			error('lftps needs "host" configured', 4);
		end

		if not config.user then
			error('lftps needs "user" configured', 4);
		end

		if not config.pass then
			error('lftps needs "pass" configured', 4);
		end

		if not config.targetdir then
			error('lftp needs "targetdir" configured', 4)
		end

		if config.target then
			error('lftp needs NOT "target" configured', 4)
		end

--		if config.exclude then
--			error('lftp does not work with excludes', 4)
--		end

		if config.rsyncOpts then
			error('lftp needs NOT "rsyncOpts" configured', 4)
		end

		if string.sub(config.targetdir, -1) == '/' then
			error('please make targetdir not end with a /', 4)
		end

	end,

	-----
	-- Default delay
	--
	delay = 1,

  	-----
	-- Exit codes for rsync.
	--
	exitcodes = {
	  [0] = 'ok',
	  [1] = 'ok',
	},
}
```

script test.lua

```lua
lftp = dofile('lftp.lua')

settings {
  nodaemon     = true,
  statusFile   = 'transfer/status',
  pidfile      = 'transfer/pid',
  inotifyMode  = 'CloseWrite',
  maxProcesses = 2,
  maxDelays    = 2,
}

sync {
	lftp,
	host      = 'sftp://fbsd-devbox',
	user      = 'eto',
	pass      = 'ssh-agent',
	source    = '/tmp/lsyncd/data',
	targetdir = '/tmp/lsyncd/data',

  exclude   = {
    'ignore.file',
  },
}
```
this is test directory layout at sender:

```text
% tree
.
├── data
│   ├── ignore.file
│   └── transfer.file
├── lftp.lua
├── lsyncd
├── test.lua
└── transfer
    ├── pid
    └── status
```

**Comment by etosan** [comment]
Do you need more data, what can I do?

**Comment by axkibe** [comment]
This is an akward product of the way Lsyncd combines configurations, by taking integer keys to the tables as "inherit" command, by default.

To fix this in your config, you'll have to add '_verbatim = true' to the exitcodes tables, so it takes the integer values as they are not trying to inherit them.

Since _verbatim should always be set for 'exitcodes' I just pushed a fix that forces it for the 'exitcodes' entry. So with the git version, it should run without having to add '_verbatim = true'.

**Comment by etosan** [comment]
Confirmed, latest git version now works with all my scripts.

Thank you very much!
